runtime/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSubscribeFactory.java

```diff
@@ -166,21 +167,26 @@ public MessageConsumer newStream(
         final long routedId = begin.routedId();
         final long initialId = begin.streamId();
         final long authorization = begin.authorization();
+        final OctetsFW extension = begin.extension();
+        final MqttBeginExFW mqttBeginEx = extension.get(mqttBeginExRO::tryWrap);
 
-        final MqttKafkaBindingConfig binding = supplyBinding.apply(routedId);
+        assert mqttBeginEx.kind() == MqttBeginExFW.KIND_SUBSCRIBE;
+        final MqttSubscribeBeginExFW mqttSubscribeBeginEx = mqttBeginEx.subscribe();
 
-        final MqttKafkaRouteConfig resolved = binding != null ?
-            binding.resolve(authorization) : null;
+        final MqttKafkaBindingConfig binding = supplyBinding.apply(routedId);
+        final List<String> topicFilters = new ArrayList<>();
+        mqttSubscribeBeginEx.filters().forEach(f -> topicFilters.add(f.pattern().asString()));
 
+        final List<MqttKafkaRouteConfig> routes = binding != null ?
+            binding.resolve(authorization, topicFilters) : null;
         MessageConsumer newStream = null;
 
-        if (resolved != null)
+        if (routes != null && !routes.isEmpty())
         {
-            final long resolvedId = resolved.id;
-            final String16FW kafkaMessagesTopic = binding.messagesTopic();
-            final String16FW kafkaRetainedTopic = binding.retainedTopic();
-            newStream = new MqttSubscribeProxy(mqtt, originId, routedId, initialId, resolvedId,
-                kafkaMessagesTopic, kafkaRetainedTopic)::onMqttMessage;
+            final String16FW defaultMessagesTopic = binding.messagesTopic();
+            final String16FW retainedTopic = binding.retainedTopic();
+            newStream = new MqttSubscribeProxy(mqtt, originId, routedId, initialId, routes,
```

Comment:
Each route should be able to answer which messages or retained topic to use, no need to implement defaulting here, that should be encapsulated inside the route config or with config implementation.